src/webcore/executor.rs

```diff
@@ -0,0 +1,242 @@
+    shrink_counter: Cell<usize>
+}
+
+// Not strictly necessary, but may become relevant in the future
+impl Drop for EventLoopInner {
+    #[inline]
+    fn drop( &mut self ) {
+        js! { @(no_return)
+            @{&self.waker}.drop();
+        }
+    }
+}
+
+impl EventLoopInner {
+    // Initializes the event loop. Only called once.
+    fn new() -> Self {
+        EventLoopInner {
+            microtask_queue: RefCell::new(VecDeque::with_capacity(INITIAL_QUEUE_CAPACITY)),
+            waker: js!(
+                var callback = @{|| EventLoop.drain()};
+                var wrapper = function() {
+                    if (!callback.dropped) { callback() }
+                };
+                var nextTick;
+
+                // Modern browsers can use `MutationObserver` which allows
+                // us to schedule a micro-task without allocating a promise.
+                // https://dom.spec.whatwg.org/#notify-mutation-observers
+                if ( typeof MutationObserver === "function" ) {
+                    var node = document.createTextNode( "0" );
+                    var state = false;
+
+                    new MutationObserver( wrapper ).observe( node, { characterData: true } );
+
+                    nextTick = function() {
+                        state = !state;
+                        node.data = ( state ? "1" : "0" );
+                    };
+
+                // Node.js and other environments
+                } else {
+                    var promise = Promise.resolve( null );
+
+                    nextTick = function() {
+                        promise.then( wrapper );
+                    };
+                }
+
+                nextTick.drop = function() {
+                    callback.dropped = true;
+                    callback.drop();
+                };
+
+                return nextTick;
+            ).try_into().unwrap(),
+            shrink_counter: Cell::new(0)
+        }
+    }
+    // Pushes a task onto the queue
+    fn push_task(&self, task: Rc< SpawnedTask >) {
+        let mut queue = self.microtask_queue.borrow_mut();
+        queue.push_back(task);
+
+        // If the queue was previously empty, then we need to schedule
+        // the queue to be drained.
+        if queue.len() == 1 {
+            self.wake();
+        }
+    }
+    // Invoke the JavaScript waker function
+    fn wake(&self) {
+        js! { @(no_return) @{&self.waker}(); }
+    }
+    // Remove and return a task from the front of the queue
+    fn pop_task(&self) -> Option< Rc< SpawnedTask > > {
+        self.microtask_queue.borrow_mut().pop_front()
+    }
+    // See if it's worth trying to reclaim some space from the queue
+    fn estimate_realloc_capacity(&self) -> Option<usize> {
+        let queue = self.microtask_queue.borrow();
+        let cap = queue.capacity();
+        // We consider shrinking the queue if it is less than
+        // half full...
+        if cap > queue.len()*2 && cap > INITIAL_QUEUE_CAPACITY {
```

Comment:
This doesn't block my approval, however [it is possible for the capacity of the `VecDeque` to be higher than `INITIAL_QUEUE_CAPACITY`](https://play.rust-lang.org/?gist=e77984289da72463397cf8a4cf6d5831&version=stable).

That means it will reallocate every 25 iterations, even if the queue never goes above `INITIAL_QUEUE_CAPACITY` elements.

Either this should be fixed, or a `TODO` note should be added.

Reply:
And... this is why [more control is needed over the capacity](https://github.com/rust-lang/rust/pull/49400) 😛 With a `shrink_to` method it wouldn't matter.

Reply:
I fixed the issue by allowing it up to 2*INITIAL_QUEUE_CAPACITY without trying to shrink (`VecDeque` uses powers of two).

Reply:
More control might not help. The `if` branch conditional will still be taken, with or without a `shrink_to` method. So you're hoping that the `VecDeque::shrink_to` implementation *always* gives exactly the same result as `VecDeque::with_capacity`, which isn't guaranteed.

Reply:
It doesn't matter what `shrink_to` gives - as long as it doesn't reallocate if the new capacity is no different from the old one.

Reply:
Also, I really don't like relying upon implementation details here. How about instead computing the capacity of the `VecDeque` immediately after it is created, and then using that instead of `INITIAL_QUEUE_CAPACITY`? That way it will work no matter what implementation `VecDeque` uses.

Reply:
That's my point: if our desired capacity is 10, and the current capacity is 15, and we call `shrink_to(10)`, if the implementation of `shrink_to` doesn't exactly match the implementation of `with_capacity`, then it might decide to reallocate it to a capacity of 14, or 16, or anything else (as long as it's greater than 10). So you're relying upon that implementation detail.

Reply:
We're not relying on implementation details for correctness, only performance, and what is performance tuning if not relying as much as possible on implementation details?

Reply:
I do not want performance to mysteriously change based upon the implementation details of `VecDeque` changing. I have explained a potential implementation which does not rely upon implementation details, so let's use that instead.

Reply:
> it might decide to reallocate it to a capacity of 14, or 16, or anything else (as long as it's >= 10). So you're relying upon that implementation detail.

We're relying on `shrink_to` not producing a random value, yes: even if it differs from `with_capacity` it will only resize once before settling down.

Reply:
Also, if it were to produce 16, then nothing would happen since the capacity would already be smaller than that.

Reply:
> I do not want performance to mysteriously change based upon the implementation details of VecDeque changing.

The implementation details of `VecDeque` are not going to change to waste more than a factor of 2 in space, and this implementation is compatible with any other change.

> so let's use that instead.

No, there's no reason to introduce another variable and code to track that to handle a situation that's not going to happen.

Reply:
Source please? My understanding is that the capacity is not guaranteed in the slightest, and it could change to use any strategy it wishes, including choosing the capacity dynamically.

Yes it's *unlikely* that it will change, but that's not the same as it never changing.

Reply:
> My understanding is that the capacity is not guaranteed in the slightest, and it could change to use any strategy it wishes, including choosing the capacity dynamically.

In which case your solution is just as broken.

As far as re-allocation strategies go, it's well known that sub-power-of-2 allocation ratios are preferred, as they reduce fragmentation when re-allocating. The main reason power of 2 is used here is because it uses bit-masking to wrap indexes, and computers use base 2.

Nothing is certain, but given those facts, I can be more confident that the allocation ratio is not going to increase than I can be in any of the other myriad of assumptions underpinning the very existence of this crate. Even if it does increase, it's not going to break anything, so the impact/probability ratio is so low that this entire conversation is irrelevant.

You simply *cannot* performance tune without relying on implementation details, and attempting to do so is a waste of time, and is the definition of premature optimisation.

Reply:
No, my solution isn't broken. If you create a `VecDeque` with capacity 10, but it's actually capacity 15, then we know if it hasn't gone above capacity 15 then it doesn't need to be reallocated, because Rust *does* guarantee that it will only reallocate if the number of elements goes above the current capacity. So it doesn't rely upon implementation details *at all*.